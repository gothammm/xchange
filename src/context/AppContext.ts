import { createContext } from 'react';
import CurrencyType from '../enums/CurrencyType';

export interface WalletType {
  currency: CurrencyType;
  value: number;
  name: string;
  isPrimary: boolean;
}

export interface AppContextType {
  wallets: WalletType[];
  update: (data: AppContextType) => {} 
}

const AppContext = createContext<AppContextType | null>(null);

export default AppContext;