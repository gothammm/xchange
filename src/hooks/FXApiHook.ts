import { useState } from 'react';
import CurrencyType from '../enums/CurrencyType';

const BASE_URL = `https://api.exchangeratesapi.io`;

interface FXResponse {
  rates: {
    [key: string]: string;
  };
  base: string;
}

interface FXResult {
  from: {
    currency: CurrencyType;
    value: number;
  };
  to: {
    currency: CurrencyType;
    value: number;
  };
  rate: number;
}

export const useFXApiHook = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState();
  const [result, setResult] = useState<FXResult>();

  return {
    error,
    result,
    isLoading,
    fx: (value: number, from: CurrencyType, to: CurrencyType) => {
      setIsLoading(true);
      return fetch(`${BASE_URL}/latest?base=${from}&symbols=${to}`)
        .then(response => response.json())
        .then((fxData: FXResponse) => {
          const conversionRate = +fxData.rates[to];
          setResult({
            from: {
              value,
              currency: from,
            },
            to: {
              value: +(value * conversionRate).toFixed(2),
              currency: to,
            },
            rate: conversionRate,
          });
        })
        .catch(e => setError(e))
        .finally(() => setIsLoading(false));
    },
  };
};
