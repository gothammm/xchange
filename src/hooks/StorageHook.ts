import { useState, useEffect } from 'react';

const STORAGE_KEY = `xchange-data`;

export const useAppStorage = () => {
  const currentData = JSON.parse(
    localStorage.getItem(STORAGE_KEY) || JSON.stringify({})
  );
  const [data, setData] = useState(currentData);

  useEffect(() => {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
  }, [data]);

  return {
    getStorageData: () => data,
    updateStorageData: (data: any) => setData(data),
  };
};
