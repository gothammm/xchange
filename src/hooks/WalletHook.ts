import { useContext } from 'react';
import AppContext, { WalletType } from '../context/AppContext';
import CurrencyType from '../enums/CurrencyType';

export const initialWalletData = [
  {
    currency: CurrencyType.USD,
    value: 0,
    name: 'USD Wallet',
    isPrimary: true,
    id: 0,
  },
  {
    currency: CurrencyType.GBP,
    value: 0,
    name: 'GBP Wallet',
    isPrimary: false,
    id: 1,
  },
  {
    currency: CurrencyType.EUR,
    value: 0,
    name: 'EUR Wallet',
    isPrimary: false,
    id: 3,
  },
];

export const useWallet = () => {
  const appContext = useContext(AppContext);
  return {
    wallets: appContext?.wallets,
    makePrimary: (wallet: WalletType) => {
      const updatedWallets =
        appContext?.wallets?.reduce(
          (acc, item) => [
            ...acc,
            { ...item, isPrimary: wallet.id === item.id },
          ],
          [] as WalletType[]
        ) || [];
      appContext?.update({ ...appContext, wallets: updatedWallets });
    },
    resetWallet: () =>
      appContext?.update({
        ...appContext,
        wallets: initialWalletData,
      }),
    addWallet: (
      name: string,
      currency: CurrencyType,
      initialValue: number = 0
    ) => {
      const newWallet = {
        name,
        currency,
        value: initialValue,
        isPrimary: false,
      } as WalletType;
      appContext?.update({
        ...appContext,
        wallets: (appContext.wallets || []).concat([newWallet]),
      });
    },
  };
};
