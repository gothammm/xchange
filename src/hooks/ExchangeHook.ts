import { WalletType } from '../context/AppContext';
import { useWallet } from './WalletHook';

export const useExchange = () => {
  const {
    addMoneyToWallet,
    deductMoneyFromWallet,
    updateWallet,
    wallets,
    commit,
  } = useWallet();

  return {
    exchange: (
      fromWallet: WalletType,
      amountToDeductFromWallet: number,
      toWallet: WalletType,
      amountToAddToWallet: number
    ) => {
      return Promise.resolve().then(() => {
        if (fromWallet.value - amountToDeductFromWallet < 0) {
          return Promise.reject(
            `Not enough money in wallet with currency ${fromWallet.currency} to exchange`
          );
        }
        const updatedContextWallets = [
          deductMoneyFromWallet(fromWallet, amountToDeductFromWallet, false),
          addMoneyToWallet(toWallet, amountToAddToWallet, false),
        ].reduce((contextWallets, wallet) => {
          return updateWallet(wallet, contextWallets);
        }, (wallets || []) as WalletType[]);

        commit(updatedContextWallets);
        return Promise.resolve();
      });
    },
  };
};
