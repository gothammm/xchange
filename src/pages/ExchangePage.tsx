import React, { useState, useEffect, useRef } from 'react';
import CenterPageContainer from '../components/CenterPageContainer';
import { useWallet } from '../hooks/WalletHook';
import CurrencyInput from '../components/CurrencyInput';
import { Empty, Button, message } from 'antd';
import { useHistory } from 'react-router-dom';
import styled from 'styled-components';
import FlexContainer from '../components/FlexContainer';
import ExchangeInformation from '../components/ExchangeInformation';
import { useFXApiHook } from '../hooks/FXApiHook';
import { WalletType } from '../context/AppContext';
import { useExchange } from '../hooks/ExchangeHook';
import WalletSelect from '../components/WalletSelect';

const InputContainer = styled(FlexContainer)`
  justify-content: flex-end;
  flex: none;
`;

const CenterContainer = styled(FlexContainer)`
  justify-content: center;
`;

const ExchangeButton = styled(Button)`
  margin-top: 2em;
`;

const ErrorInWallet: React.FC<{ message: string; onHomeClick: () => void }> = ({
  message,
  onHomeClick,
}) => {
  return (
    <CenterPageContainer>
      <Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description={message}>
        <Button type="primary" onClick={onHomeClick}>
          Go Home
        </Button>
      </Empty>
    </CenterPageContainer>
  );
};

const ExchangePage: React.FC = () => {
  const { getPrimaryWallet, wallets } = useWallet();
  const { exchange } = useExchange();
  const history = useHistory();
  if (!wallets) {
    return (
      <ErrorInWallet
        onHomeClick={() => history.push('/')}
        message={`Data was cleared, no wallet found.`}
      />
    );
  }
  const activeWallet = getPrimaryWallet();
  const [showWallets, setShowWallets] = useState(false);

  const [hasSwapped, setHasSwapped] = useState(false);
  const otherWallets =
    (activeWallet &&
      wallets?.filter(wallet => wallet.id !== activeWallet.id)) ||
    [];
  const [secondaryWallet, setSecondaryWallet] = useState(otherWallets[0]);
  const [isExchanging, setIsExchanging] = useState(false);
  const [walletToChange, setWalletToChange] = useState<WalletType | null>(null);
  const fx = useFXApiHook();
  const [exchangeValue, setExchangeValue] = useState<{
    from: number;
    to: number;
  }>({ from: 0, to: 0 });

  if (!wallets) {
    return (
      <ErrorInWallet
        onHomeClick={() => history.push('/')}
        message={`Data was cleared, no wallet found.`}
      />
    );
  }

  if (!activeWallet) {
    return (
      <ErrorInWallet
        onHomeClick={() => history.push('/')}
        message={`Oops, no primary wallet was found.`}
      />
    );
  }
  const [primaryWallet, setPrimaryWallet] = useState(activeWallet);

  if (otherWallets && otherWallets.length <= 0) {
    return (
      <ErrorInWallet
        onHomeClick={() => history.push('/')}
        message={`Oops, only primary wallet was found, we require more than one wallet for exchange.`}
      />
    );
  }

  const fetchFXRate = async (primary: WalletType, secondary: WalletType) => {
    return fx.getFXRate(primary.currency, secondary.currency);
  };

  const onWalletChange = (wallet: WalletType) => {
    setWalletToChange(wallet);
    setShowWallets(true);
  };

  useEffect(() => {
    const interval = setInterval(() => {
      fetchFXRate(primaryWallet, secondaryWallet);
      message.info(`Exchange rate has been refreshed`);
    }, 10000);
    fetchFXRate(primaryWallet, secondaryWallet);
    return () => clearInterval(interval);
  }, []);

  useEffect(() => {
    fetchFXRate(primaryWallet, secondaryWallet);
  }, [secondaryWallet.id, primaryWallet.id]);

  useEffect(() => {
    const from = exchangeValue.from;
    const to = exchangeValue.to;
    if (from <= 0 && to <= 0) return;
    if (!hasSwapped) return;
    if (fx.error || !fx.result) {
      message.error(`FX Rate data is unavailable, try again.`);
      return;
    }
    const doReverse = activeWallet.currency !== fx.result.from;
    if (primaryWallet.id !== activeWallet.id) {
      return setExchangeValue({
        to: from,
        from: fx.getFXValue(fx.result, from, doReverse),
      });
    }
    if (primaryWallet.id === activeWallet.id) {
      return setExchangeValue({
        from: to,
        to: fx.getFXValue(fx.result, to, doReverse),
      });
    }
  }, [primaryWallet.id, secondaryWallet.id]);

  return (
    <CenterPageContainer>
      <CenterContainer>
        <InputContainer>
          <CurrencyInput
            onWalletChange={onWalletChange}
            wallet={primaryWallet}
            onChange={value => {
              setExchangeValue({
                from: value,
                to:
                  !fx.error && fx.result
                    ? fx.getFXValue(fx.result, value)
                    : value,
              });
            }}
            value={exchangeValue.from}
          />
        </InputContainer>
        <ExchangeInformation
          isLoading={fx.isLoading}
          conversionRate={fx.result}
          primary={primaryWallet}
          secondary={secondaryWallet}
          onRefreshRate={() => fetchFXRate(primaryWallet, secondaryWallet)}
          onSwap={(newPrimary, newSecondary) => {
            setPrimaryWallet(newPrimary);
            setSecondaryWallet(newSecondary);
            fetchFXRate(newPrimary, newSecondary);
            !hasSwapped && setHasSwapped(true);
          }}
        />
        <InputContainer>
          <CurrencyInput
            onWalletChange={onWalletChange}
            wallet={secondaryWallet}
            onChange={value => {
              setExchangeValue({
                to: value,
                from:
                  !fx.error && fx.result
                    ? fx.getFXValue(fx.result, value)
                    : value,
              });
            }}
            value={exchangeValue.to}
          />
        </InputContainer>
        <ExchangeButton
          loading={isExchanging}
          disabled={exchangeValue.from <= 0 && exchangeValue.to <= 0}
          type="primary"
          shape="round"
          onClick={() => {
            setIsExchanging(true);
            exchange(
              primaryWallet,
              exchangeValue.from / 100,
              secondaryWallet,
              exchangeValue.to / 100
            )
              .then(() => {
                message.success(`Exchange successful!`);
                setIsExchanging(false);
                history.push('/');
              })
              .catch(e => {
                setIsExchanging(false);
                message.error(e);
              });
          }}
        >
          Exchange
        </ExchangeButton>
      </CenterContainer>
      <WalletSelect
        onWalletSelect={(wallet: WalletType) => {
          if (!walletToChange || !wallet) {
            return;
          }
          if (walletToChange.id === wallet.id) return;
          if (walletToChange.id === primaryWallet.id) {
            setPrimaryWallet(wallet);
            secondaryWallet.id === wallet.id &&
              setSecondaryWallet(primaryWallet);
            fetchFXRate(wallet, primaryWallet);
          } else if (walletToChange.id === secondaryWallet.id) {
            setSecondaryWallet(wallet);
            primaryWallet.id === wallet.id && setPrimaryWallet(secondaryWallet);
            fetchFXRate(secondaryWallet, wallet);
          }
          setShowWallets(false);
        }}
        show={showWallets}
        onClose={() => setShowWallets(false)}
        wallets={
          wallets.map(wallet =>
            walletToChange && walletToChange.id === wallet.id
              ? { ...wallet, isPrimary: true }
              : { ...wallet, isPrimary: false }
          ) || []
        }
        text={`Available Wallets / Currency`}
      />
    </CenterPageContainer>
  );
};

export default ExchangePage;
