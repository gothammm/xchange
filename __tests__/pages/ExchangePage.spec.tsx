import React, { ReactNode } from 'react';
import { render, fireEvent } from '@testing-library/react';
import ExchangePage from '../../src/pages/ExchangePage';
import { MemoryRouter } from 'react-router-dom';
import AppContext from '../../src/context/AppContext';
import { useAppStorage } from '../../src/hooks/StorageHook';
import { renderHook, act } from '@testing-library/react-hooks';
import { useWallet } from '../../src/hooks/WalletHook';
import fetch from 'jest-fetch-mock';

describe(`ExchangePage component`, () => {
  let appWrapper: React.ComponentType;
  const exchangeRate = 1.3;
  beforeEach(() => {
    localStorage.clear();
    appWrapper = ({ children }: { children?: ReactNode }) => {
      const { updateStorageData, getStorageData } = useAppStorage();
      return (
        <AppContext.Provider
          value={{
            ...getStorageData(),
            update: updateStorageData,
          }}
        >
          {children}
        </AppContext.Provider>
      );
    };
    fetch.mockResponse(req => {
      if (req.url.includes('/latest')) {
        return Promise.resolve({
          body: JSON.stringify({
            rates: {
              GBP: exchangeRate.toString(),
              EUR: exchangeRate.toString(),
            },
            base: 'USD',
          }),
          status: 200,
        });
      }
      return Promise.reject({
        status: 400,
        body: `error`,
      });
    });
  });

  test(`should match snapshot`, () => {
    const component = render(
      <MemoryRouter>
        <ExchangePage />
      </MemoryRouter>
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test(`should convert currency between two currency input`, async () => {
    const { result } = renderHook(() => useWallet(), { wrapper: appWrapper });
    const initialMoney = 100;
    act(() => {
      result.current.wallets?.forEach(wallet => {
        result.current?.addMoneyToWallet(wallet, initialMoney);
      });
    });
    const component = render(
      <MemoryRouter>
        <ExchangePage />
      </MemoryRouter>,
      { wrapper: appWrapper }
    );
    const usdInput: HTMLInputElement = (await component.findByTestId(
      'USD'
    )) as HTMLInputElement;
    fireEvent.keyDown(usdInput, { key: '1', code: 10 });
    fireEvent.keyDown(usdInput, { key: '0', code: 10 });
    fireEvent.keyDown(usdInput, { key: '0', code: 10 });
    fireEvent.keyDown(usdInput, { key: '0', code: 10 });
    fireEvent.keyDown(usdInput, { key: '0', code: 10 });

    expect(usdInput.value).toEqual('$100.00');

    const allInputs: HTMLInputElement[] = (await component.findAllByRole(
      'currency-input'
    )) as HTMLInputElement[];
    const exchangeInput: HTMLInputElement | null =
      allInputs.find(
        item => item.attributes?.getNamedItem('data-testid')?.value !== 'USD'
      ) || null;
    expect(exchangeInput).toBeTruthy();
    expect(exchangeInput?.value).toEqual('Â£130.00');
  });
});
