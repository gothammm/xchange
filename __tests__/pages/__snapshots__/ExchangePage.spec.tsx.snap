// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangePage component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm sc-AxirZ NiafK"
  >
    <div
      class="sc-AxjAm sc-fzoLsD bExCFN"
    >
      <div
        class="sc-AxjAm sc-fzozJi bMpsPL"
      >
        <div
          class="sc-AxhUy jXvmIW"
        >
          <span
            class="ant-input-affix-wrapper"
          >
            <span
              class="ant-input-prefix"
            >
              <div
                class="sc-AxiKw dIfukv"
              >
                <div
                  class="sc-AxhCb gIsOZG"
                  data-testid="USD:switch-wallet"
                >
                  <h3
                    class="ant-typography"
                  >
                    USD
                  </h3>
                  <span
                    aria-label="down"
                    class="anticon anticon-down"
                    role="img"
                    style="margin-left: 5px; font-size: 15px;"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </div>
                <span
                  class="ant-typography ant-typography-secondary"
                >
                  Balance: $0.00
                </span>
              </div>
            </span>
            <input
              class="ant-input"
              data-testid="USD"
              placeholder="Enter value"
              role="currency-input"
              type="text"
              value="- $0.00"
            />
          </span>
        </div>
      </div>
      <div
        class="sc-AxjAm sc-AxgMl hcHxKw"
      >
        <button
          class="ant-btn ant-btn-primary ant-btn-circle ant-btn-icon-only"
          type="button"
        >
          <span
            aria-label="swap"
            class="anticon anticon-swap"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="swap"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M847.9 592H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h605.2L612.9 851c-4.1 5.2-.4 13 6.3 13h72.5c4.9 0 9.5-2.2 12.6-6.1l168.8-214.1c16.5-21 1.6-51.8-25.2-51.8zM872 356H266.8l144.3-183c4.1-5.2.4-13-6.3-13h-72.5c-4.9 0-9.5 2.2-12.6 6.1L150.9 380.2c-16.5 21-1.6 51.8 25.1 51.8h696c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary ant-btn-round ant-btn-icon-only ant-btn-loading"
          type="button"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading"
            role="img"
          >
            <svg
              aria-hidden="true"
              class="anticon-spin"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="sc-AxjAm sc-fzozJi bMpsPL"
      >
        <div
          class="sc-AxhUy jXvmIW"
        >
          <span
            class="ant-input-affix-wrapper"
          >
            <span
              class="ant-input-prefix"
            >
              <div
                class="sc-AxiKw dIfukv"
              >
                <div
                  class="sc-AxhCb gIsOZG"
                  data-testid="GBP:switch-wallet"
                >
                  <h3
                    class="ant-typography"
                  >
                    GBP
                  </h3>
                  <span
                    aria-label="down"
                    class="anticon anticon-down"
                    role="img"
                    style="margin-left: 5px; font-size: 15px;"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </div>
                <span
                  class="ant-typography ant-typography-secondary"
                >
                  Balance: £0.00
                </span>
              </div>
            </span>
            <input
              class="ant-input"
              data-testid="GBP"
              placeholder="Enter value"
              role="currency-input"
              type="text"
              value="+ £0.00"
            />
          </span>
        </div>
      </div>
      <button
        class="ant-btn sc-fzpans jQyOhp ant-btn-primary ant-btn-round"
        disabled=""
        type="button"
      >
        <span>
          Exchange
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
